### PR TITLE
MESOS: don't attempt to schedule static pods

### PR DESCRIPTION
**Pull request by jdef**
TODO:
- [x] add static pod to mesos-docker cluster for testing
- [x] restructure static pod config/zipper code to be more amenable to munging
- [x] modify static pod configs to include the proper binding annotation
- [x] switch from zip archive to gzip'd `api.PodList` for static pods
- [x] write the host binding annotation in the executor, not the scheduler; clean up procurement!
- [x] unit tests
- [x] e2e test for static pods

xref https://github.com/mesosphere/kubernetes-mesos/issues/493

was BLOCKED by:
- #17054 :: properly deal with resource request/limit during resource procurement

### CODE REVIEW

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 1]
```diff
@@ -470,6 +433,72 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
...
+		}
+
+		cpu := mresource.PodCPULimit(pod)
+		mem := mresource.PodMemLimit(pod)
+		log.V(2).Infof("reserving %.2f cpu shares and %.2f MB of memory to static pod %s/%s", cpu, mem, pod.Namespace, pod.Name)
+
+		staticPodCPUs += float64(cpu)
+		staticPodMem += float64(mem)
+
+		// TODO(jdef) add binding annotation so that static pods aren't picked up by the scheduler
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: this is where we'd add the missing binding annotation

File: contrib/mesos/pkg/scheduler/service/service_test.go [hunk 2]
```diff
@@ -124,42 +124,3 @@ func Test_DefaultResourceLimits(t *testing.T) {
 	assert.Equal(s.DefaultContainerCPULimit, mresource.DefaultDefaultContainerCPULimit)
 	assert.Equal(s.DefaultContainerMemLimit, mresource.DefaultDefaultContainerMemLimit)
 }
-
-func Test_StaticPods(t *testing.T) {
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service_test.go]: this test was redundant, so I removed it

**Comment by jdef** [comment]
@sttts why did you delete this branch?

**Comment by sttts** [comment]
Oops. Sorry. I thought that's the old static-pods implementation branch.

**Comment by sttts** [comment]
We should think about some branch naming scheme.

**Comment by jdef** [comment]
agreed. for example, if it's tied one of our k8sm issues, perhaps something
like `k8sm-{issueId}-{shortName}`

**Comment by sttts** [comment]
[k8sm-{issueid}]-{feature|cleanup|bug}-{shortName}

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 3]
```diff
@@ -60,13 +59,15 @@ import (
...
-	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/resource"
 	client "k8s.io/kubernetes/pkg/client/unversioned"
 	clientauth "k8s.io/kubernetes/pkg/client/unversioned/auth"
 	"k8s.io/kubernetes/pkg/master/ports"
 	etcdstorage "k8s.io/kubernetes/pkg/storage/etcd"
 	"k8s.io/kubernetes/pkg/tools"
+
+	// lock to this API version, compilation will fail when this becomes unsupported
+	_ "k8s.io/kubernetes/pkg/api/v1"
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: should probably invoke an empty/noop func in v1 to really force the compiler to pick up on this (endpoints controller used to do something similar with the old v1beta1, v1beta2)

**Comment by jdef** [comment]
@sttts @s-urbaniak PTAL (yes, I still need to write unit tests)

File: cluster/mesos/docker/docker-compose.yml [hunk 4]
```diff
@@ -158,6 +158,7 @@ scheduler:
     --mesos-master=mesosmaster1:5050
     --cluster-dns=10.10.10.10
     --cluster-domain=cluster.local
+    --static-pods-config=/static-pods
```
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: Before merging we should remove this commit again.

File: contrib/mesos/pkg/scheduler/service/util.go [hunk 5]
```diff
@@ -86,3 +94,131 @@ func hashExecutorInfo(info *mesos.ExecutorInfo) uint64 {
...
+		applyBindingsTo := filterFunc(func(pod *api.Pod) (bool, error) {
+			if pod.Annotations == nil {
+				pod.Annotations = make(map[string]string)
+			}
+			// add binding annotation so that static pods aren't picked up by the main scheduling loop.
+			// static pods are always completely managed by the kubelet.
+			pod.Annotations[meta.BindingHostKey] = assignedSlave
+			return true, nil
+		})
+		return archive.Create(applyBindingsTo.podEntries(archive.Playback(entries)))
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: Can we get rid of the archiving code? Just make a pod list, marshal it into json and gzip it. 100x simpler. No code to maintain.

File: contrib/mesos/pkg/scheduler/service/util.go [hunk 6]
```diff
@@ -86,3 +94,131 @@ func hashExecutorInfo(info *mesos.ExecutorInfo) uint64 {
...
+				}
+				entry.Data = updated
+				out <- entry
+			}
+		}
+	}()
+	return out
+}
+
+func makeStaticPodValidator(limitCPU resource.CPUShares, limitMem resource.MegaBytes, accumCPU, accumMem *float64, minimalResources bool) filterFunc {
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: Having this code in util.go is not nice. Quoting Tomas: put code into util.go is a sign of bad concepts. At least move it into its own file staticpods.go.

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 7]
```diff
@@ -666,13 +657,16 @@ func (s *SchedulerServer) bootstrap(hks hyperkube.Interface, sc *schedcfg.Config
 
 	as := scheduler.NewAllocationStrategy(
 		podtask.DefaultPredicate,
-		podtask.NewDefaultProcurement(s.DefaultContainerCPULimit, s.DefaultContainerMemLimit))
+		podtask.NewDefaultProcurement(
+			s.DefaultContainerCPULimit,
+			s.DefaultContainerMemLimit,
+			staticPodProcurement))
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service.go]: This is nice.

File: contrib/mesos/pkg/scheduler/service/service.go [hunk 8]
```diff
@@ -412,49 +413,7 @@ func (s *SchedulerServer) prepareExecutorInfo(hks hyperkube.Interface) (*mesos.E
...
-			log.V(2).Infof("reserving %.2f cpu shares and %.2f MB of memory to static pod %s", cpu, mem, pod.Name)
-
-			staticPodCPUs += float64(cpu)
-			staticPodMem += float64(mem)
-		}
-
-		// pass zipped pod spec to executor
-		execInfo.Data = bs
-	}
+	bs, staticPodCPUs, staticPodMem := s.prepareStaticPods()
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service.go]: Why bs and not spp or something readable?

**Comment by sttts** [comment]
For my taste this implementation is far too complex. I would like to see how the following approach turns out:
- get rid of the zip archive, we use compressed json instead
- parse the pod files and store the pod list as simple `[]*api.Pod` in the procurement
- in the procurement clone the pod list, set the annotation, marshal it to json, compress and set as data of the ExecutorInfo
- try to keep the static pod logic compact at one place. Now it's spread over multiple files, e.g. the binding code is in util.go, not in procurement.

Using simple pod lists feels much more direct than juggling with zip archives. We don't support directory structures anyway, only a flat file list.

File: contrib/mesos/pkg/scheduler/service/util.go (hunk 5 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/util.go]: The file source doesn't support pod lists, only single pod-per-file. And at
the moment we're trying to re-use the existing pod sources vs. creating new
ones.

File: contrib/mesos/pkg/scheduler/service/service.go (hunk 8 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service.go]: will rename

**Comment by jdef** [comment]
one alternative is to write our own file pod config source. i was trying to
avoid heading down this path.

another alternative is to leverage the kubelet's http-source for static
pods and serve them from the scheduler. this is less complex from an
initial deployment standpoint since the kubelet then becomes responsible
for monitoring changes to the static pod configuration and, in addition,
injecting the binding annotation. the downside is that the scheduler has no
way to account for these resources upon launch of the executor: it could do
an initial estimate based on a scan of the static pod config dir, but then
if anything in the dir changes (new/removed pods; resource limit changes)
there's then a race between the scheduler somehow obtaining additional
resources for the static pod and the executor processing the changes (of
course you could get around this in the scheduler by cloning the initial
static pods dir into a separate dir whose lifecycle is controlled by the
scheduler -- and http-serve the static pod configs from there .. ugh).

your point re: flat vs. nested directory structures is a good one. not sure
why the initial implementation recurses a filesystem vs. simply listing the
files in a directory. we could refactor that.

the current implementation is also not flexible w/ respect to *changes* to
static pod config after the initial deployment: pretty sure that if we
change the executorInfo.data field and try to launch a task on a slave that
already has an executor running that the mesos-master will slap our hands.
static pods also seem pretty hackish in the wake of the recently introduced
daemon-controller upstream (glorified replication controller, maintains
copies of 'daemon pod' specs across every node). i suspect that once
daemon-controlled pods are in the non-experimental, stable API that we'll
be able to kill support for static pods in our implementation (unless
there's a specific conformance test that requires that we have them).

i'm in agreement that, overall, this code seems complex given our goal.
haven't seen/thought of something that i like more.

File: contrib/mesos/pkg/scheduler/service/util.go (hunk 6 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/util.go]: ok

File: contrib/mesos/pkg/scheduler/service/util.go (hunk 5 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: I don't say to use another file source. We can still write the json snippets out into flat files. But for the transport and the patching we don't need a zip. The file source also does not support directories. So there is no need for this.

File: contrib/mesos/pkg/scheduler/service/service_test.go (hunk 2 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service_test.go]: there is another one for that?

Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/service_test.go]: this isn't doing anything static-pod specific, it's just testing the archive package. the archive package already has unit tests

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/service_test.go]: True. I think in the LifeCycle test we actually have a real test case.

File: contrib/mesos/pkg/scheduler/service/util.go (hunk 5 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/util.go]: If we did the pod list thing I'd be tempted to capture the file name in an
annotation so that when we create the static pods dir on the executor, the
layout makes sense to someone trying to debug why their static pod may not
be working as expected (not our fault, hopefully!). But then what's the
difference between a compressed list of pod entries that know their file
names and a zip archive? The complexity feels almost the same to me. The
zip archive adds complexity in that it supports a multi-layer filesystem
structure (that we don't really need) --- but that's somewhat trivial in
the code; and we also add an extra deserialize/serialize step since the
pods in the zip container are stored in byte form (we only pay this price
once per slave).

I guess I'm questioning whether it's worth duplicating some of the
functionality that zip gives us by managing a pod-list.

Will think on this more.

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: If we use the pod list, we can easily do the binding modification in the executor instead of pulling this into the procurement.

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: And then we only pay the prize of marshaling once per run, not even once per slave.

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: And it will be natural to change the medium one day to get the specs to the kubelet.

Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/util.go]: that's a very good point

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/util.go]: Btw, I like the idea of annotating the pod with the file name.

**Comment by jdef** [comment]
squashed commits then rebased to master (shippable wasn't building)

File: contrib/mesos/pkg/executor/executor_test.go [hunk 9]
```diff
@@ -254,15 +248,13 @@ func TestExecutorLaunchAndKillTask(t *testing.T) {
 	mockDriver.AssertExpectations(t)
 }
 
+/*
```
Comment (jdef) [inline contrib/mesos/pkg/executor/executor_test.go]: need to restore/update this unit test

File: contrib/mesos/pkg/staticpods/gzip_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
+/*
```
Comment (sttts) [inline contrib/mesos/pkg/staticpods/gzip_test.go]: staticpods/doc.go is missing.

File: contrib/mesos/pkg/staticpods/gzip.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+func GZip(pods <-chan *api.Pod) ([]byte, error) {
+	return gzipList(List(pods))
+}
+
+func gzipList(list *api.PodList) ([]byte, error) {
```
Comment (sttts) [inline contrib/mesos/pkg/staticpods/gzip.go]: nit: if it's GZip, it should also be gZip....

File: contrib/mesos/pkg/staticpods/io.go [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+			}
+			annotate(&pod.ObjectMeta, map[string]string{meta.StaticPodFilename: f.Name()})
+			pods <- pod
+		}
+	}()
+	return pods, errors
+}
+
+// tryDecodeSinglePod was copied from pkg/kubelet/config/common.go v1.0.5
+func tryDecodeSinglePod(data []byte) (parsed bool, pod *api.Pod, err error) {
```
Comment (sttts) [inline contrib/mesos/pkg/staticpods/io.go]: Would it be an option to make the kubelet version public instead of copying?

Comment (jdef) [inline contrib/mesos/pkg/staticpods/io.go]: maybe. the kubelet version is a little more complicated than this is now.
there's really not a lot of magic in this, so it's probably not worth
keeping the xref comment.

File: contrib/mesos/pkg/staticpods/gzip.go (hunk 11 above)
Comment (jdef) [inline contrib/mesos/pkg/staticpods/gzip.go]: ugh. i would hate to type gZip. i had settled on gzip and GZip because they
felt better. I'd rather change GZip to Gzip if we want absolute consistency

File: contrib/mesos/pkg/staticpods/io.go (hunk 12 above)
Comment (sttts) [inline contrib/mesos/pkg/staticpods/io.go]: only an idea. If it doesn't make sense, this is fine.

File: contrib/mesos/pkg/staticpods/gzip.go (hunk 11 above)
Comment (sttts) [inline contrib/mesos/pkg/staticpods/gzip.go]: I am not pedantic about naming. There are much more important things in code ;-) How you like.

**Comment by sttts** [comment]
lgtm. After squashing :rocket:

**Comment by sttts** [comment]
Don't forget to remove the first commit.

**Comment by sttts** [comment]
Though for the sake of keeping this stable, I wouldn't mind to leave the static pod there. But I would move it into the staticpods namespace. And please add a e2e test to test that it's up on each node.

**Comment by jdef** [comment]
squashed and rebased

**Comment by sttts** [comment]
Is this ready for review?

**Comment by jdef** [comment]
I was waiting until I had an e2e test before asking for final review

**Comment by jdef** [comment]
e2e testing is currently broken .. waiting for it to be fixed before I can verify the test that I wrote

**Comment by jdef** [comment]
rebased

File: contrib/mesos/pkg/executor/executor.go [hunk 13]
```diff
@@ -280,13 +293,12 @@ func (k *KubernetesExecutor) Reregistered(driver bindings.ExecutorDriver, slaveI
...
-func (k *KubernetesExecutor) initializeStaticPodsSource(data []byte) {
+func (k *KubernetesExecutor) initializeStaticPodsSource(hostname string, data []byte) error {
 	log.V(2).Infof("extracting static pods config to %s", k.staticPodsConfigPath)
-	err := archive.UnzipDir(data, k.staticPodsConfigPath)
-	if err != nil {
-		log.Errorf("Failed to extract static pod config: %v", err)
-		return
-	}
+	annotator := podutil.Annotator(map[string]string{
+		meta.BindingHostKey: hostname,
```
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Please comment why meta.BindingHostKey is set

File: contrib/mesos/pkg/scheduler/service/validation.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+// a CPU or Memory limit. if minimalResources is true then podtask.MinimalCpus and
+// podtask.MinimalMem are used when accounting for pod resources (but are not applied as pod
+// resource limits). minimalResources is part of a bigger hack/work-around that attempts
+// to address a systemd/mesos conflict w/ respect to resource accounting.
+func staticPodValidator(accumCPU, accumMem *float64, minimalResources bool) podutil.FilterFunc {
+	return podutil.FilterFunc(func(pod *api.Pod) (bool, error) {
+		// TODO(sttts): allow unlimited static pods as well and patch in the default resource limits
+		// for now use of the value 1 is a placeholder: we're discarding pods if they have containers
+		// without limits so it doesn't really matter what number we use here.
+		unlimitedCPU := resource.LimitPodCPU(pod, 1)
```
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/validation.go]: should wait for the changes on #17054 so that we don't end up with pods that have requested resources >= the limit that we calculate here

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/validation.go]: Can you rebase this one on-top of #17054?

Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/validation.go]: I've named that PR as a blocker for this. Going through the process of
rebasing is busy-work since that PR will be merged very soon (it's already
lgtm'd). I'd rather just wait for it to go through and then rebase to
master.

Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/validation.go]: when the submit queue is unblocked again....

**Comment by jdef** [comment]
I signed it!
.. along with sttts

**Comment by jdef** [comment]
TC smoke test failure:
```
[12:15:46][Step 5/5] Error pulling image (v2.2.1) from quay.io/coreos/etcd, ApplyLayer exit status 1 stdout:  stderr: write /etcd: no space left on device
[12:15:46][Step 5/5] 4adf2b3afd05: Error downloading dependent layers
[12:15:46][Step 5/5] ea3ced1122e1: Error pulling image (v2.2.1) from quay.io/coreos/etcd, endpoint: https://quay.io/v1/, ApplyLayer exit status 1 stdout:  stderr: write /etcd: no space left on device
[12:15:46][Step 5/5] ea3ced1122e1: Error pulling image (v2.2.1) from quay.io/coreos/etcd, ApplyLayer exit status 1 stdout:  stderr: write /etcd: no space left on device
```

File: cluster/mesos/docker/deploy-addons.sh [hunk 15]
```diff
@@ -54,6 +54,7 @@ function deploy_ui {
 
 # create the kube-system namespace
```
Comment (sttts) [inline cluster/mesos/docker/deploy-addons.sh]: nit: fix comment

File: cluster/mesos/docker/docker-compose.yml [hunk 16]
```diff
@@ -148,6 +149,8 @@ scheduler:
   - etcd
   - mesosmaster1
   - apiserver
+  volumes:
+  - ./static-pod.json:/static-pods/static-pod.json
```
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: nit: shouldn't we map that into /opt as well?

File: contrib/mesos/pkg/executor/executor.go [hunk 17]
```diff
@@ -206,10 +206,23 @@ func (k *Executor) isDone() bool {
...
 
 // sendPodUpdate assumes that caller is holding state lock; returns true when update is sent otherwise false
-func (k *Executor) sendPodUpdate(u *kubetypes.PodUpdate) bool {
+func (k *Executor) sendPodUpdate(sendSnapshot bool) bool {
 	if k.isDone() {
 		return false
 	}
+	var snapshot []*api.Pod
+	if sendSnapshot {
+		snapshot = make([]*api.Pod, 0, len(k.pods))
```
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: This function feels strange. It used to send an update as the name suggests. Why was this changed? You can call the old function with 
```
&kubetypes.PodUpdate{
		Op:   kubetypes.SET,
		Pods: []*api.Pod{},
	}
```

File: contrib/mesos/pkg/executor/executor.go [hunk 18]
```diff
@@ -240,10 +256,7 @@ func (k *Executor) Registered(driver bindings.ExecutorDriver,
 	// emit an empty update to allow the mesos "source" to be marked as seen
 	k.lock.Lock()
 	defer k.lock.Unlock()
-	k.sendPodUpdate(&kubetypes.PodUpdate{
-		Pods: []*api.Pod{},
-		Op:   kubetypes.SET,
-	})
+	k.sendPodUpdate(false)
```
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: As written above. Don't like this change.

Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Isn't k.pods empty here anyway? Is this bool  needed?

File: contrib/mesos/pkg/executor/service/podsource.go [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+// sourceMesos merges pods from mesos, and mirror pods from the apiserver. why?
+// (a) can't have two sources with the same name;
+// (b) all sources, other than ApiserverSource are considered static/mirror
+// sources, and;
+// (c) kubelet wants to see mirror pods reflected in a non-static source.
+//
+// Mesos pods must appear to come from apiserver due to (b), while reflected
+// static pods (mirror pods) must appear to come from apiserver due to (c).
+//
```
Comment (sttts) [inline contrib/mesos/pkg/executor/service/podsource.go]: :+1:  for the docs

File: contrib/mesos/pkg/executor/service/podsource.go [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+		execUpdates:    execUpdates,
+		out:            out,
+	}
+	// reflect changes from the watch into a chan, filtered to include only mirror pods (have an ConfigMirrorAnnotationKey attr)
+	cache.NewReflector(podWatch, &api.Pod{}, cache.NewUndeltaStore(source.send, cache.MetaNamespaceKeyFunc), 0).RunUntil(sourceFinished)
+	go source.mergeAndForward()
+}
+
+func (source *sourceMesos) send(objs []interface{}) {
+	var pods []*api.Pod
```
Comment (sttts) [inline contrib/mesos/pkg/executor/service/podsource.go]: nit: pods -> mirrorPods

File: contrib/mesos/pkg/executor/service/podsource.go [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+	var (
+		mirrors = []*api.Pod{}
+		pods    = []*api.Pod{}
+	)
+eventLoop:
+	for {
+		select {
+		case m := <-source.mirrorPods:
+			mirrors = make([]*api.Pod, len(m))
+			copy(mirrors, m)
```
Comment (sttts) [inline contrib/mesos/pkg/executor/service/podsource.go]: Why copy? Can't we assume that the receiver becomes the owner of the slice?

File: contrib/mesos/pkg/podutil/filters.go [hunk 22]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+func Annotator(m map[string]string) FilterFunc {
+	return FilterFunc(func(pod *api.Pod) (bool, error) {
+		Annotate(&pod.ObjectMeta, m)
+		return true, nil
+	})
+}
+
+func Stream(list *api.PodList, err error) <-chan *api.Pod {
```
Comment (sttts) [inline contrib/mesos/pkg/podutil/filters.go]: add comment please

File: contrib/mesos/pkg/podutil/filters.go [hunk 23]
```diff
@@ -0,0 +1,85 @@
...
+	//have much to put there right now so it can just live here.
+	if meta.Annotations == nil {
+		meta.Annotations = make(map[string]string)
+	}
+	for k, v := range kv {
+		meta.Annotations[k] = v
+	}
+}
+
+func Annotator(m map[string]string) FilterFunc {
```
Comment (sttts) [inline contrib/mesos/pkg/podutil/filters.go]: add comment please

File: contrib/mesos/pkg/podutil/filters.go [hunk 24]
```diff
@@ -0,0 +1,85 @@
...
+	"k8s.io/kubernetes/pkg/api"
+)
+
+type defaultFunc func(pod *api.Pod) error
+
+type FilterFunc func(pod *api.Pod) (bool, error)
+
+type Filters []FilterFunc
+
+func Annotate(meta *api.ObjectMeta, kv map[string]string) {
```
Comment (sttts) [inline contrib/mesos/pkg/podutil/filters.go]: add comment please

File: contrib/mesos/pkg/podutil/filters.go [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+				log.Errorf("pod failed selection: %v", err)
+			} else if ok {
+				out <- pod
+			}
+		}
+	}()
+	return out
+}
+
+func List(pods <-chan *api.Pod) *api.PodList {
```
Comment (sttts) [inline contrib/mesos/pkg/podutil/filters.go]: add comment please

File: contrib/mesos/pkg/scheduler/service/validation_test.go [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+		assert.Equal(t, tc.podcount, len(list.Items), "test case #%d: expected %d pods instead of %d", i, tc.podcount, len(list.Items))
+		assert.EqualValues(t, tc.cputot, cpu, "test case #%d: expected %f total cpu instead of %f", i, tc.cputot, cpu)
+		assert.EqualValues(t, tc.memtot, mem, "test case #%d: expected %f total mem instead of %f", i, tc.memtot, mem)
+	}
+}
+
+type podOpt func(*api.Pod)
+type ctOpt func(*api.Container)
+
+func _pods(pods ...*api.Pod) <-chan *api.Pod {
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/validation_test.go]: why everything with underscore?

File: test/e2e/mesos.go [hunk 27]
```diff
@@ -50,4 +53,16 @@ var _ = Describe("Mesos", func() {
 		}
 		Expect(len(addr)).NotTo(Equal(""))
 	})
+
+	It("starts static pods on every node in the mesos cluster", func() {
+		client := framework.Client
+		expectNoError(allNodesReady(client, 2*time.Minute), "all nodes ready")
```
Comment (sttts) [inline test/e2e/mesos.go]: Please use pkg.util.ForeverTestTimeout instead of hardcoded timeouts.

File: test/e2e/mesos.go [hunk 28]
```diff
@@ -50,4 +53,16 @@ var _ = Describe("Mesos", func() {
...
+
+		nodelist, err := client.Nodes().List(labels.Everything(), fields.Everything())
+		expectNoError(err, "nodes fetched from apiserver")
+
+		ns, numpods := "static-pods", len(nodelist.Items)
+		expectNoError(waitForPodsRunningReady(ns, numpods, 5*time.Minute),
```
Comment (sttts) [inline test/e2e/mesos.go]: Please use pkg.util.ForeverTestTimeout instead of hardcoded timeouts.

File: test/e2e/mesos.go [hunk 29]
```diff
@@ -50,4 +53,16 @@ var _ = Describe("Mesos", func() {
...
 	})
+
+	It("starts static pods on every node in the mesos cluster", func() {
+		client := framework.Client
+		expectNoError(allNodesReady(client, 2*time.Minute), "all nodes ready")
+
+		nodelist, err := client.Nodes().List(labels.Everything(), fields.Everything())
+		expectNoError(err, "nodes fetched from apiserver")
+
+		ns, numpods := "static-pods", len(nodelist.Items)
```
Comment (sttts) [inline test/e2e/mesos.go]: Two lines are much easier to read.

File: contrib/mesos/pkg/scheduler/podtask/minimal.go [hunk 30]
```diff
@@ -24,25 +24,27 @@ import (
...
+)
 
-	DefaultMinimalProcurement = AllOrNothingProcurement([]Procurement{
+func NewDefaultMinimalProcurement(other ...Procurement) Procurement {
+	return Procurement(AllOrNothingProcurement(append([]Procurement{
 		ValidateProcurement,
 		NodeProcurement,
 		MinimalPodResourcesProcurement,
 		PortsProcurement,
-	}).Procure
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/podtask/minimal.go]: The changes here in this file are not used at all, aren't they?

File: contrib/mesos/pkg/scheduler/meta/annotations.go [hunk 31]
```diff
@@ -30,4 +30,5 @@ const (
 	PortNameMappingKeyPrefix = "k8s.mesosphere.io/portName_"
 	PortNameMappingKeyFormat = PortNameMappingKeyPrefix + "%s_%s"
 	ContainerPortKeyFormat   = "k8s.mesosphere.io/containerPort_%s_%s_%d"
+	StaticPodFilename        = "k8s.mesosphere.io/staticPodFilename"
```
Comment (sttts) [inline contrib/mesos/pkg/scheduler/meta/annotations.go]: StaticPodFilename -> StaticPodFilenameKey

File: contrib/mesos/pkg/podutil/doc.go [hunk 32]
```diff
@@ -14,6 +14,6 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package archive provides utilities to archive and unarchive filesystem
-// hierarchies.
-package archive
+// podutil contains utilities for reading, writing, and filtering streams
```
Comment (sttts) [inline contrib/mesos/pkg/podutil/doc.go]: nit: no , behind writing

File: contrib/mesos/pkg/executor/service/podsource.go [hunk 33]
```diff
@@ -0,0 +1,132 @@
+/*
```
Comment (sttts) [inline contrib/mesos/pkg/executor/service/podsource.go]: For the sake of using the type system for private/public checking we can move this source into contrib/mesos/pkg/executor/source.

File: contrib/mesos/pkg/executor/service/service.go [hunk 34]
```diff
@@ -111,10 +111,8 @@ func (s *KubeletExecutorServer) runExecutor(execUpdates chan<- kubetypes.PodUpda
 			return status, nil
 		},
 		StaticPodsConfigPath: staticPodsConfigPath,
-		PodLW: cache.NewListWatchFromClient(apiclient, "pods", api.NamespaceAll,
-			fields.OneTermEqualSelector(client.PodHost, s.HostnameOverride),
-		),
-		NodeInfos: nodeInfos,
+		PodLW:                podWatch,
```
Comment (sttts) [inline contrib/mesos/pkg/executor/service/service.go]: use the same identifiers, either podLW or PodWatch.

File: cluster/mesos/docker/static-pod.json [hunk 35]
```diff
@@ -0,0 +1,29 @@
...
+    "namespace": "static-pods"
+  },
+  "spec": {
+    "containers": [
+      {
+        "image": "busybox",
+        "resources": {
+          "limits": {
+            "cpu": "100m",
+            "memory": "256Mi"
```
Comment (sttts) [inline cluster/mesos/docker/static-pod.json]: 256 Mi is probably far too much. How about 64?

**Comment by sttts** [comment]
Overall it looks pretty good. A number of bigger and smaller nits.

File: cluster/mesos/docker/docker-compose.yml (hunk 16 above)
Comment (jdef) [inline cluster/mesos/docker/docker-compose.yml]: the scheduler wants a path to a directory that contains pod config files. i'm happy to re-parent /static-pods under /opt.

File: contrib/mesos/pkg/executor/executor.go (hunk 17 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: we send snapshots (SET operations) because that what the new merged source deals with. I didn't feel like copy/pasting logic everywhere to convert the pods map into a slice, so I centralized that here. I can rename the func to something like `sendPodsSnapshot` if you think that's more clear.

File: contrib/mesos/pkg/executor/executor.go (hunk 18 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: it should be. maybe I was being paranoid when I wrote this?>

File: cluster/mesos/docker/docker-compose.yml (hunk 16 above)
Comment (sttts) [inline cluster/mesos/docker/docker-compose.yml]: would be more consistent.

File: contrib/mesos/pkg/executor/service/podsource.go (hunk 21 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/service/podsource.go]: I was debugging something and decided to `copy()` instead of grabbing a shallow copy (via `m[:]`)

File: contrib/mesos/pkg/executor/executor.go (hunk 17 above)
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: sendPodsSnapshot without the bool would be perfect, if it works without that.

File: contrib/mesos/pkg/scheduler/service/validation_test.go (hunk 26 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/service/validation_test.go]: because I wanted short, easily readable names that I could be certain wouldn't be mistaken for anything other than test funcs. i'll rename if it hurts.

File: contrib/mesos/pkg/scheduler/podtask/minimal.go (hunk 30 above)
Comment (jdef) [inline contrib/mesos/pkg/scheduler/podtask/minimal.go]: hm .. i don't think they are, any more. i can clean up

File: contrib/mesos/pkg/executor/service/podsource.go (hunk 33 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/service/podsource.go]: right, we talked about this. i wasn't crazy about moving this into its own package simply for type private/public checking -- i'm not expecting anyone to abuse this, and it's ugly hack code which is why I think it belongs in /service

File: contrib/mesos/pkg/scheduler/service/validation_test.go (hunk 26 above)
Comment (sttts) [inline contrib/mesos/pkg/scheduler/service/validation_test.go]: It hurts a little. I am not pedantic about those things though. What about using another package to avoid name conflicts?

**Comment by jdef** [comment]
@k8s-bot please test this

**Comment by sttts** [comment]
@k8s-bot test this please

File: contrib/mesos/pkg/executor/executor.go [hunk 36]
```diff
@@ -390,10 +400,7 @@ func (k *Executor) handleChangedApiserverPod(pod *api.Pod) {
 			oldPod.DeletionTimestamp = pod.DeletionTimestamp
 			oldPod.DeletionGracePeriodSeconds = pod.DeletionGracePeriodSeconds
 
-			k.sendPodUpdate(&kubetypes.PodUpdate{
-				Op:   kubetypes.UPDATE,
-				Pods: []*api.Pod{oldPod},
-			})
+			k.sendPodsSnapshot()
```
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Can you add a comment that we send the first birth cry here with an empty pod list?

**Comment by sttts** [comment]
Small nit. Otherwise, looks good to me.

File: contrib/mesos/pkg/executor/executor.go (hunk 36 above)
Comment (jdef) [inline contrib/mesos/pkg/executor/executor.go]: this is more like a death-cry. is your comment misplaced?

Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: Oh right. Didn't see the context in the diff.

File: contrib/mesos/pkg/executor/executor.go [hunk 37]
```diff
@@ -240,10 +251,7 @@ func (k *Executor) Registered(driver bindings.ExecutorDriver,
 	// emit an empty update to allow the mesos "source" to be marked as seen
```
Comment (sttts) [inline contrib/mesos/pkg/executor/executor.go]: here we have the birth cry :+1:
